### DISCUSSION TITLE
param-rewrite-with-matching does not longer compile with v3.3.19

### DISCUSSION THREAD

**Discussion by tueddy** [mathieucarbou/ESPAsyncWebServer#143] [General]
I have param-rewrite code exactly like in documention:
https://github.com/mathieucarbou/ESPAsyncWebServer?tab=readme-ov-file#param-rewrite-with-matching

With 3.3.19 the code no longer compiles:

`error: 'bool OneParamRewrite::match(const AsyncWebServerRequest*)' marked 'override', but does not override
  218 |     bool match(const AsyncWebServerRequest* request) override {
      |          ^~~~~
`

Can you show the new match() function and update the docs?
Thanks!

**Comment by mathieucarbou** [comment]
Hello, I did some fixes yesterday around type hierarchy - maybe I messed up something.

~~Could you please show me more of your code: which class you are extending ? This is a rewrite ?~~

**Comment by mathieucarbou** [comment]
Ok got it!

The previous API had an issue: match method should be const:

`bool match(AsyncWebServerRequest* request) const override`

Node the `const`.

This method should not modify internal states in theory....

**Comment by tueddy** [comment]
Code is exactly like in doc:

```
class OneParamRewrite : public AsyncWebRewrite {
protected:
	String _urlPrefix;
	int _paramIndex;
	String _paramsBackup;

public:
	OneParamRewrite(const char *from, const char *to)
		: AsyncWebRewrite(from, to) {

		_paramIndex = _from.indexOf('{');

		if (_paramIndex >= 0 && _from.endsWith("}")) {
			_urlPrefix = _from.substring(0, _paramIndex);
			int index = _params.indexOf('{');
			if (index >= 0) {
				_params = _params.substring(0, index);
			}
		} else {
			_urlPrefix = _from;
		}
		_paramsBackup = _params;
	}

    bool match(const AsyncWebServerRequest* request) override {
		if (request->url().startsWith(_urlPrefix)) {
			if (_paramIndex >= 0) {
				_params = _paramsBackup + request->url().substring(_paramIndex);
			} else {
				_params = _paramsBackup;
			}
			return true;

		} else {
			return false;
		}
	}
};
```
and later:

```
		wServer.addRewrite(new OneParamRewrite("/rfid/ids-only", "/rfid?ids-only=true"));
		wServer.addRewrite(new OneParamRewrite("/rfid/{id}", "/rfid?id={id}"));
```

`bool match(const AsyncWebServerRequest* request) override`  does not longer compile..

**Reply by mathieucarbou** [reply]
~~yes like I said above and if you change to add `const` it won't either because some internal states are modified in the object.~~

~~How your class is handling concurrent requests anyway with the shared state ? ~~
~~Did you benchmark it to  send concurrently a flow of requests using 16 workers ?~~
~~What is happening to the `_params ` ?~~

_edit_

Just saw that the base class has some states too... I will rollback the const! That was a mistake sorry!

**Comment by mathieucarbou** [comment]
Moved to issue https://github.com/mathieucarbou/ESPAsyncWebServer/issues/144 (this is a bug)

**Comment by tueddy** [comment]
I do not use `_params ` in any other part of code. I just picked up the doc example code to add a simple rewrite.

Calling "/rfid/ids-only" rewrites to "/rfid?ids-only=true" perfectly, so i can use both REST calling conventions.

Maybe there is an easier way adding a rewrite without declaring a new class?

**Reply by mathieucarbou** [reply]
For now I will just rollback the change I did: the new version `3.3.20` is already there, you can test it and let me know.

In theory the match function should not modify any internal state but it does and the base class works the same way...
So I will review that in the future...
The name is poorly chosen - there should probably me something else handling the parameter rewrite, and instead of the state put in the filter, have the states put in the request.

**Reply by mathieucarbou** [reply]
v3.3.20 released !
